web/api/v1/api.go

```diff
@@ -1526,7 +1529,42 @@ func (api *API) cleanTombstones(r *http.Request) apiFuncResult {
 	return apiFuncResult{nil, nil, nil, nil}
 }
 
-func (api *API) respond(w http.ResponseWriter, data interface{}, warnings storage.Warnings) {
+type arrowWriter interface {
+	WriteArrow(io.Writer) error
+	ArrowEncodeable() bool
+}
+
+func (api *API) tryArrowResponse(w http.ResponseWriter, r *http.Request, data interface{}, warnings storage.Warnings) error {
+	result, ok := data.(*queryData)
+	if !ok {
+		return fmt.Errorf("invalid result data type: %T", data)
+	}
+
+	arrowWriter, ok := result.Result.(arrowWriter)
+	if !ok {
+		return errors.Errorf("arrow not implemented for %s", result.ResultType)
+	}
+
+	if !arrowWriter.ArrowEncodeable() {
+		return fmt.Errorf("could not encode this result with arrow")
+	}
+
+	w.Header().Set("Content-Type", mimeTypeArrowStream)
+	w.WriteHeader(http.StatusOK)
+	if err := arrowWriter.WriteArrow(w); err != nil {
+		level.Error(api.logger).Log("msg", "error writing arrow response", "err", err)
+	}
+	return nil
+}
+
+func (api *API) respond(w http.ResponseWriter, r *http.Request, data interface{}, warnings storage.Warnings) {
+	if r.Header.Get("Accept") == mimeTypeArrowStream {
+		err := api.tryArrowResponse(w, r, data, warnings)
```

Comment:
Why return full error messages and errors at all if nothing is going to be done with them?

Reply:
Oops, I meant to log them at a debug level. Added the log now.

Reply:
Cool thanks